RFD 0162: Machine ID Fleet Management, `BotInstance` resource and  other `token` joining improvements

strideynet opened the pull request:
Proposes solutions for https://github.com/gravitational/teleport/issues/26885

RFD to improve the UX of `token` joining and managements of large fleets of `tbot`. Introduces a BotInstance resource to allow these to be managed and monitored in the UI and CLI - similar to how we handle heartbeats for agents today.

strideynet committed:
skeleton out rfd

strideynet committed:
Restrucutre to highlight the selected options over alternatives

strideynet committed:
identify edge cases

strideynet committed:
Explain backend key format

strideynet committed:
Add spec for heartbeat rpc

strideynet committed:
Add API for BotInstance resource

strideynet committed:
Add notes on analytics

strideynet committed:
Add a brief summary of out of scope improvements

strideynet committed:
Add output/services to heartbeat

strideynet committed:
Remove services/outputs

strideynet committed:
Expand on `tctl` command changes

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+  of this work to the `token` join method.
+- Add support for calling the join RPCs with a client certificate.
+
+This technique could reuse a recently proposed LoginID attribute. This would 
```
I'm curious about whether we'll move forward with the proposed idea of LoginIDs from the hackathon - we could even straight up re-use this for join/renews of Machine ID.

strideynet replied:
It looks like there's not really been much movement on this since January. We can probably safely assume we do not want to depend on that.

timothyb89 replied:
Since it looks like we're dropping this, I'll remove the reference.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+// Request for GetBotInstance.
+message GetBotInstanceRequest {
+  // The name of the BotInstance to retrieve.
+  string name = 1;
```
I wonder if it makes more sense to seperate the bot name and fingerprint fields here rather than provide them concatenated.

strideynet replied:
Yeah -  I think I'm growing to like the idea of explicit fields - but I do see one potential issue. If I already have a BotInstance resource, it's very easy to call GetBotInstance using the name of the resource in this field if I want to  check for any updates. If we switch to separate fields, there's potentially a little more complexity in this as you'd have to do something like

```
GetBotInstance({
  Bot: instance.spec.bot,
  Instance: instance.name
})
```

This feels fairly minor so I haven't got a super strong opinion here. Feel free to change.

timothyb89 replied:
I've split the fields, though on some level I wonder if we could just refer directly to the bot instance identifier. The bot name namespaces the UUID, but is that really necessary?

strideynet replied:
Hm - but with only the UUID, we won't be able to get the BotInstance from the store, as we're going to include the bot name in the storage key to enable listing of BotInstances for a given bot.

timothyb89 replied:
Ah, good call, didn't think about that.

strideynet commented:
@timothyb89 I did some thinking over the weekend and I stumbled across a fairly big decision I think we need to make.

Should we either:

- Maintain the existing behaviour of delegated joining today, where we essentially "rejoin" for each renewal and the generation counter mechanism is not used.
- Switch to using the renewal/generation counter mechanism for all join methods

Thoughts on maintaining existing behaviour:

- Each BotInstance will have multiple joins associated with it. This could make how we display information a little more complicated (e.g if we wanted to show what GitHub action a BotInstance is associated with). Whilst we expect each of these joins/delegated identities to be similar/identical, we can't prevent someone switching join method for an already joined bot.
- Avoids relying more on the currently flakey generation counter.
- If a user deletes the join token, already joined bots will no longer be able to renew.
- Avoids a change from the behaviour users are used to today.

Thoughts on switching to renewal mechanism:

- Each BotInstance has a single initial join associated with it. This makes attributing actions to a specific delegated identity simpler.
- If a user deletes the join token, already joined bots can continue to renew and function as expected. We could add an option when deleting a token that then "locks out" any botinstances that joined using that token to essentially mimic the previous behaviour. This is more similar to how join tokens work for agent joining, as they do not currently need to renew, so deleting their join token does not impact them.
- Generation counter is fairly flakey at the moment, we'd need to make this more stable.
- Makes behaviour consistent across join methods and bots vs agents. Today, the `token` join method is "special" as the only renewable join method. This means it's inconsistent with the other join methods and how they work, and also means that the other join methods behave differently when used for a bot and when used for an agent.

For me, I think the key point to this is the UX around deleting a token: is it better for the bots to then fail to renew or for them to continue to function?

thedevelopnik commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+Bot instance and encode this within the certificate. Upon renewals, the UUID
+would be copied from the current certificate and into the new one.
+
+Whilst this is fairly easy to implement for the `token` join method, one
```
This limitation would go away if we moved the mechanism for delegated joins over to this mechanism, correct?

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+fingerprint of the public key. In addition, the full public key should be
+recorded and verified against when authenticating a Bot instance action.
+
+It should be noted that with this technique, rotating the keypair of a `tbot`
+instance would reset the identity of that instance. Rotation of this keypair
+would be unusual and this side effect seems expected.
```
> We do create fresh keys for impersonated identities on each renewal today.

I lean towards this being slightly different - the impersonated identities go to an output which by it's very nature will be less secure than the bot's own internal storage. You may have a number of consumers that pull these secrets out into different places.

timothyb89 replied:
Agreed, though while it's certainly _less_ pressing to rotate bots' internal keypairs, it's arguably a bit weird that we just never do it today. In other words, I think private key rotation is a reasonable thing to implement in the future, and some dedicated identifier would let us maintain a single `BotInstance` through that.

On the other hand, the negative consequence to rotating if we used public key identifiers would just be a new bot instance. It's convenient to have a clear lineage, but probably not critical.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+previous certificate of the Bot instance is not readily available. We can
+either:
+
+- Accept this limitation and treat each renewal of a delegated Bot instance as a
+  new Bot instance. This is likely unacceptable and would limit any advantages
+  of this work to the `token` join method.
```
> I'd expect all delegated joins from CI providers to be new bot instances every time, since they're stateless.

Hmm, I'm not sure I'd necessarily want to set that precedent. Even in CI situations, a long-running CI job may renew several times (e.g a background tbot database tunnel with a CI job running a long migration or an ansible job connecting to hundreds of hosts to roll out an update) and I think the events associated with that long-running job being associated with different bot instances would be confusing. 

> And I imagine a decent share of the cloud provider / k8s joins that could be stateful end up being ephemeral anyway.

Right now, there's not much of a motivation to have state configured because there's no real advantage to it, in fact, at one point there was a disadvantage as an expired cert would cause the bot to be unable to start. I'd imagine that if we were to introduce BotInstance's then there's a perceived benefit at least in being able to attribute actions to a particular instance of tbot throughout its lifetime in a fairly persistent way.

I suppose the other overriding thought in my head is I'm desperate to stop the token vs delegated join method behaviour diverting further. It's already confusing enough and I can see user frustration when the less secure join method is actually delivering more features to them (in terms of having a long-lived persistence identity they can attribute actions to).

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+  // is reached. This avoids the indefinite growth of the resource but also
+  // ensures the initial record is retained.
+  repeated BotInstanceStatusAuthentication authentications = 4;
+  // Last X records kept, with the second oldest being removed once the limit
+  // is reached. This avoids the indefinite growth of the resource but also
+  // ensures the initial record is retained.
+  repeated BotInstanceStatusHeartbeat heartbeats = 5;
```
Sure - feel free to make this change. I don't really have a strong opinion about how this is stored.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+
+### Changes to the `token` Join Method
+
+As we now have a way to track the generation for a specific Bot instance, we
+can allow multiple Bot instances to be associated with a single Bot. This
+also means that the token no longer needs to be consumed on a join.
+
+Eventually, we may wish to add a way to specify a number of joins which can
+occur with a token. This provides a way to easily control the lifetime of a 
+token when deploying to a fleet of a pre-known size.
+
+The renewal logic will need to be adjusted to read and update the generation
+counter from the BotInstance rather than the Bot user.
```
> I'm a bit conflicted about launching multi-use tokens without also adding options to limit joins.

I agree some sort of limiting factor on a number of joins is likely going to be necessary out of the box.

> We should include a join count limit from the start, and arguably default it to 1 so users are making an explicit decision to allow multiple bots to join. This default would also avoid a potentially surprising change in behavior for existing users.

I like this idea - especially using a default of 1 to avoid a big change in behaviour.

---

One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

We'd still want them to rotate though... one thing I was thinking about to make this easier would be to produce a script or command which can be executed on a cron job that generates a new token for a named bot and then calls a user-supplied script which should connect to their CI platform/secret store and updates the token. This may be a nice answer for those who are backed into a corner by a rudimentary CI platform.

This fits well into the "bot refresh" idea, some sort of easyish CLI command where you specify an existing bot and can just generate a new token.

timothyb89 replied:
> One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

I agree. I almost wonder if there's a middle ground here where we might be able to force users to make a security decision:
* The `token` join method sets a sane cap on the TTL of a join token to proactively encourage some variety of automation, say 7 days or something. I don't think we need to cap the number of uses as long as there's a time bound.
  * It might be useful to write up a cron job example, maybe a guide on using a GHA or k8s cronjob to rotate a Jenkins joining secret or something.
* We then (eventually) introduce a new `token-insecure` that otherwise behaves identically but has no restrictions at all on TTL or number of joins (basically this [other thread](https://github.com/gravitational/teleport/pull/36510/files#r1602429025) down the page). This method might also opt out of the generation counter.

I think we could argue enforcing a maximum token TTL is an acceptable breaking change given how narrowly useful long TTL'd tokens are today, and then introduce `token-insecure` at some point in the future? I can't imagine any users today are making practical use of extremely long TTL bot tokens that would have their use cases broken.

strideynet replied:
I'm a little cautious about setting a cap on the TTL, if we were totally greenfield I'd agree but:
- Today, there's no cap. I agree with you and doubt anyone is relying on this behaviour - so perhaps we can discount this.
- We'd need to have logic that ignores the cap if it's for Agent joining
- We'd need to have logic that ignores the cap if the join method is not `token`.

I'd perhaps be less cautious about the TTL cap if we were approaching this at a later date when we add `token-insecure`, imho, the whole token type could do with a V3 and that might be a good opportunity to do so. As an aside, I'd love to make it so token name is not the actual secret value, e.g a `token` join method token has a UUID-esque name or a user provided name, and then a field which specifies the actual secret value. This would make audit logging etc much easier because we wouldn't need to "hide" the name.

I'm not overly decided on this and I think I need to ponder on it a little longer to figure out my full thoughts. It's almost out of scope for this RFD - so perhaps we don't need to make a decision on this immediately and we can proceed with the BotInstance work since I don't think this has any impact.

timothyb89 replied:
I think my hopefully more in scope recommendation for this RFD is: let's enforce a soft client-side limit in `tctl` and/or warn that it's a bad idea and we may restrict token length in the future. I figure that if a user wants to create an extremely long-lived token, they can write up the YAML themselves and `tctl create -f` it.

I think warnings or soft errors today help keep the door open for us to lock this down more in the future while making sure use cases that need very long lived tokens are at least possible.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+This token would be incredibly sensitive and if stolen, there would be no
+automated mechanisms to detect this as exists today with the generation counter.
+
+It likely makes more sense to improve the existing `token` join method rather
+than introduce a variant which behaves differently and is less secure. It would
+increase the complexity of the codebase and the user experience.
```
Definitely out of scope for this quarter - but well worth investigating I should think.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+We could introduce a multi-phase commit of the generation counter. This would
+provide more robustness to the renewal process.
+
+### Locking of Individual Bot Instances
+
+Currently, it's only possible to lock out an entire Bot user. This means that
+when managing a large fleet, it would not be able to lock out a specific host
+that had been compromised. This is likely to be a major friction point for those
+deploying a large number of Bot instances.
+
+It also increases the significance of the fragility of the generation counter.
```
One thing I thought about here is an option to "Lock all bot instances that have authenticated with X token". Seems fairly feasible if we store the hash of the token used on the first join or similar.

I think this is deffo something for next quarter though.

timothyb89 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,565 @@
+one-to-one binding between a single instance of `tbot` and a single Bot user.
+This is not the case when using the delegated join methods.
+ 
+### Persistent Bot Instance Identity
```
After some experimentation, I think we'll actually need to do most of the more complex work involved in UUID verification even if we pick public key identifiers.

The problem, I think, is that `RegisterUsingTokenRequest` doesn't check the public key regardless, so clients could just lie about it. This means a malicious bot could potentially impersonate an existing instance. I don't think that's a hugely critical vulnerability but might help a malicious bot disguise itself.

I think we'll want to pass along an existing certificate to verify regardless... and we have to verify private key ownership, so that means we still need to return the new certs encrypted with the provided public key, after verifying it against auth's CA.

With that in mind, I don't think it's much of a stretch to include additional fields in the cert if we want.

It might honestly also be worthwhile to look at adding a grpc entrypoint to delegated joining. We'd then get a proper TLS session with the existing client cert which is IMO much less sketchy than manually verifying the certs and re-encrypting the payload from the existing endpoint.

strideynet replied:
Yeah - that's a good point. Although, even if we pursue the simple model of encrypting the returned certs, for the purposes of the audit logs, a malicious actor can still pretend to be another bot instance. I thought about requiring a signature when making the initial request but I worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

As you say, a gRPC entrypoint would resolve most of these concerns and would allow for any challenge and response flows we wanted. It'd also be an easy way to flip between the old and new behaviour in regards to BotInstances. I think the only challenge worth bearing in mind with a gRPC entrypoint is how an unauthenticated client would reach the auth server, today, the IAM and Azure join methods are terminated at the Proxy and the gRPC requests are re-made, but this would conceal the TLS certificate from the auth server. Perhaps that's not really a problem and unauthenticated requests could be made to the proxy directly, and when you have an existing cert, you can tunnel through and hit the auth server directly.

I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep. There's a lot in the RegisterUsingToken flow that people will want to refactor and change, and we'll need to resist that if we want to avoid this becoming a rather large task. Perhaps we set out a bounds and say for the V1 of the gRPC method, we'll mimic the HTTP fields precisely.

timothyb89 replied:
>  worry about how those can relatively easily be re-sent unless the signed value is some sort of challenge - perhaps use a timestamp ???

Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

> I like the idea generally and moving away from HTTP and inline with other efforts - my only major concern with moving this to gRPC would be avoiding scope creep

:+1: I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

To cut scope initially I'd potentially be okay with a gRPC endpoint that is _only_ accessible by already authenticated users. Clients would onboard through HTTP, but renew via gRPC. I'll experiment and see what makes sense, if anything.

strideynet replied:
> Yep, I tried to game out what a proper challenge would look like and was not liking the amount of effort or complexity it would add given it still feels like a hack vs a gRPC registration entrypoint.

Yeah the only thing that comes to mind is signing a timestamp and having the auth verify that it falls within a 5 minute window - but that's not fantastic for a bunch of reasons.

> 👍 I'll pivot my investigation today to gauge how complex a gRPC refresh would be.

Sounds good - gRPC gives us the option to do call and responses which is great. The TPM joining I did fairly recently is a nice-ish example of how to achieve this and does show how to use the existing JoinUsingTokenRequest proto within a challenge-response flow.

timothyb89 replied:
Alright, I'll be trying to merge our discussions back into the RFD today. My summary so far is:
* We'd ideally like to use UUIDs (or similar) as the primary bot identifier, to allow (among other things) key rotation; I'll rewrite this section to list that as the preferred method and public keys as the alternate
* I'll also add a short section on trust of this new identifier to help outline why I think it's important to verify it during renewal.

timothyb89 committed:
Update RFD with revised persistent identity and other changes

This adds a new section on the security requirements for the
persistent bot identity, alongside an updated preference for UUID
identifiers.

This also includes revisions to the "changes to the `token` join
method" to include join count limits by default and set sane
defaults to encourage short lived tokens.

Lastly, various other small changes were applied, like removing
the implementation details for bot instance names in favor of an
explicitly specified bot name and instance identifier.

timothyb89 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+message BotInstanceStatus {
+  // The unique identifier for this bot.
+  string id = 1;
+  // The public key of the Bot instance.
+  // When authenticating a Bot instance, the full public key must be compared
+  // rather than just the fingerprint to mitigate pre-image attacks.
+  bytes public_key = 2;
+  // The fingerprint of the public key of the Bot instance.
+  string fingerprint = 3;
```
I've so far kept the public key and fingerprint fields even though we're adding a UUID. Do we want to keep these? I don't think they're strictly required, but may still be useful.

strideynet replied:
I think it's useful info but if it's placed at the top level, we don't have a context point for when the value was this. We could move this into `BotInstanceStatusAuthentication` and that way we'll properly handle the key changing.

timothyb89 replied:
Good call, will move it into that message.

timothyb89 commented:
I've done a pass to apply some of the discussions we've more or less found a consensus on, mainly switching to UUID identifiers and lightly touching on how we'll verify them. I've also suggested some tweaks to the `token` join method to include join counter limits by default and discourage (but still ultimately allow) very long TTLs.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+from a CI workflow, their certificates could expire and prevent them from
+being identified as the same instance. This is likely to only impact a small
+number of cases, however, as most CI provider joins are stateless and have no
+certificates to present anyway. Bots that present expired certificates will
```
I imagine the TLS handshake will reject them for us - we can just make sure the client is capable of handling this and will try without a certificate.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+instead the endpoint will check:
+
+- The presented client certificate is for the Bot linked to the instance.
+- The presented client certificate's public key matches the public key recorded
```
We can drop this requirement now that we're using UUIDs.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+Upon each join and renewal, the BotInstance record will be updated with an
+additional entry in the `status.authentications` field. If there is X entries,
+then the second-oldest entry will be removed. This prevents growth without
+bounds but also ensures that the original record is retained.
```
Now that we're storing the initial as it's own field, we will want to delete the oldest rather than second oldest.

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+then the second-oldest entry will be removed. This prevents growth without
+bounds but also ensures that the original record is retained.
+
+In addition, the TTL of the BotInstance resource will be extended.
```
```suggestion
In addition, the TTL of the BotInstance resource will be extended to cover the validity period of the issued certificate, plus a short additional time to allow for some imprecision.
```

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,655 @@
+// Request for GetBotInstance.
+message GetBotInstanceRequest {
+  // The name of the bot associated with the instance.
+  string name = 1;
```
```suggestion
  string bot_name = 1;
```

timothyb89 committed:
Apply code review feedback

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into rfd/0162-machine-id-token-join-method-bot-instance

strideynet reviewed (commented):
I raised this PR initially but Tim has taken it over - so I can't Approve it. Please take this as my "approval".

timothyb89 committed:
Use consistent field names for authentications and heartbeats

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+- Bot: An identity within Teleport intended for use by machines as opposed to
+  humans. Many individual machines may act as this shared identity.
+- `tbot`: The Teleport binary that acts as aBot and generates credentials for
```
```suggestion
- `tbot`: The Teleport binary that acts as a Bot and generates credentials for
```

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+## Why
+
+Whilst deploying a large fleet of Bots is fairly trivial when using the
```
I might mention TPM joining here as well. If TPMs are available, that flow is probably preferred.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+The following burdens currently exist:
+
+- When using the `token` join method, a Bot must be created for each host. This
```
```suggestion
- When using the `token` join method, a Bot must be created for each Bot instance. This
```

To reuse the terminology you recently introduces.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+- When using the `token` join method, a token can only be used once. This means
+  creating hundreds of join tokens and managing securely distributing these to
+  hosts.
+- When managing a large fleet of `tbot` deployments, there is no way to
```
```suggestion
- When managing a large fleet of Bot instances, there is no way to
```

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+   decrypt the returned identity if they actually have the private key for the
+   previous identity.
+
+Our preference is option (1): bots always join over gRPC, and provide a client
+certificate when re-joining. Bots without a client cert to present are
+registered as new instances, while bots with a valid client cert preserve their
+identity.
```
+1, nice and simple

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+of BotInstances for a given Bot.
+
+Like agent heartbeats, the BotInstance will expire after a period of inactivity.
+This avoids the accumulation of ephemeral BotInstances.
```
Do we anticipate large numbers of bot instances at scale? Do we need to do something like variable rate heartbeats here, or is a fixed TTL reasonable?

timothyb89 replied:
I'd guess a customer will eventually want to create a huge number of bot instances. What do you mean by variable rate, would jitter qualify?

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+An audit event should be added for the deletion of a BotInstance. The
+name of the BotInstance should be added to the existing join, renewal and
+certificate generation audit events.
```
We should also put the bot instance's name/UUID in any existing audit events (like certificates issued) so that we have traceability into the actions of a particular bot.

In the future, we can then extend the access monitoring feature or even offer a bot details page in the UI that lists its renewals, join attempts, certs issued, etc.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+Commands to list all BotInstances and the BotInstances for a specific Bot should
+be added to the `tctl bots` family:
+
+`tctl bots instances list`
```
Do we need `instances`? What about `tctl bots ls`?

timothyb89 replied:
I think the two are distinct, `tctl bots ls` will continue to work as it does today, showing (effectively) a name + allowed roles. You'd then list all the unique joins of a given bot with `tctl bots instances list --bot <name>`.

There's some overlap, I think we could amend `tctl bots ls` to say, show a unique count of instances per bot, for example.

strideynet replied:
Yeah - that's correct, we'll still want way to list the actual bot identities.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+import "teleport/header/v1/metadata.proto";
+
+// A BotInstance
+message BotInstance {
```
Does this resource have any impact on our Terraform provider?

timothyb89 replied:
I don't think so, bot instances will be created automatically when a bot joins. They're more for record keeping and visibility than something end users need to configure.

I could see exposing it in the future if we add some command and control functionality to `BotInstanceSpec`, though.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+If the heartbeat fails, then `tbot` should retry on an exponential backoff.
+
+##### Alternative: Submit Heartbeat Data on Join/Renew
```
Before merging this, it would be helpful if each `Alternative` that we present but ultimately do not select lists why it was not chosen.

timothyb89 replied:
It's been long enough now that I don't remember the order of events, but we now have a summary of why alternatives were not selected for (at least) public key fingerprints and the heartbeat RPC.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+long TTLs will need to be explicitly specified.
+
+We may additionally want to put hurdles in the way of things like extremely long
+token TTLs. There are legitimate low-security use cases for these, but we could
```
I'm curious what Sasha thinks about this.

Extremely long join token TTLs basically become a shared secret like an API key, and we built Machine ID to avoid long-lived shared secrets..

timothyb89 replied:
I agree, and would ideally like to only support this by way of an explicitly insecure `token-insecure` join method at some point in the future (see previous thread: https://github.com/gravitational/teleport/pull/36510#discussion_r1604165929). As we discussed in that thread, adding a TTL limit would _technically_ be a breaking change, and there are use cases we may still want to support.

For context on why, today's status quo is problematic because it seems like we properly support token joining, but it's kind of an illusion. It works great in a Getting Started guide on your dev box, and then breaks as soon as you try it in production because one of your nodes went offline for more than an hour. Production use really requires full automation (e.g. ansible or terraform detecting dead bots and reissuing tokens), and I'm not aware of any customers who've actually put in the effort to do it.

The changes proposed here (multiple uses per token, multiple tokens per bot) simplify the automation requirements a lot, but it's still a heavy lift, and I'm not sure it'll ever be worthwhile for a lot of use cases. There may be value in giving customers in this niche a way to still make use of Teleport if they (explicitly) accept the security implications of a long lived token; they still benefit from our RBAC, audit logging, etc.

roraback replied:
Michael is working on the display and creation of join tokens in the web UI, and we actually _just_ had a conversation with Sasha about token creation during the Design and Product Office Hours yesterday. In the meeting, we agreed to show a danger state for any `token` join method with an expiration >24 hours. We also agreed to make it impossible to create long-lived tokens through the web UI. 

(What we discussed doesn't affect the underlying mechanics of the `token` join method, but it does put some guard rails for their creation in the web UI and affects how they will be _displayed,_ regardless of how they were created.)

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+simply clear out any artifacts within the `tbot` storage directory.
+
+In addition, if the bot detects a change in join token or join method, it should
+automatically rotate it's keypair. This will ensure it presents as a fresh
```
```suggestion
automatically rotate its keypair. This will ensure it presents as a fresh
```

But you want to do this in a way that it can still use its existing client certs to re-join, right?

timothyb89 replied:
Yes, and we'll be supporting rejoin-then-rotate regardless. I think since we record the token and join method as part of the last <n> authentication records, it makes sense for that to be our behavior if bot config changes.

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+
+### Analytics
+
+A PostHog event should be emitted for each BotInstance heartbeat. This will
+allow us to track active bots in a similar way to how we track active agents.
```
Should we included it in the aggregated `tp.resource.heartbeat` event rather than creating a new event for each heartbeat?

zmb3 commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,653 @@
+### Locking of Individual Bot Instances
+
+Currently, it's only possible to lock out an entire Bot user. This means that
+when managing a large fleet, it would not be able to lock out a specific host
+that had been compromised. This is likely to be a major friction point for those
```
We can leave this out of scope but we should explore it in the future.

We can already lock any heartbeating agent based on its Host ID today, so it feels like this should extend to bots without too much trouble.

timothyb89 replied:
Yeah, we should add _something_ here, but it's trivial to circumvent when the join token can just be reused, @strideynet had some ideas here: https://github.com/gravitational/teleport/pull/36510/files/61e17365cac7aaec059de86d470a57a6b612eebd#r1602421886

timothyb89 committed:
Apply suggestions from code review

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

timothyb89 committed:
Address review feedback: mention TPM joining, alternatives, audit

strideynet commented on rfd/0162-machine-id-token-join-method-bot-instance.md:
```diff
@@ -0,0 +1,664 @@
+---
+authors: Noah Stride (noah.stride@goteleport.com)
+state: draft
```
```suggestion
state: Implemented (16.2.0)
```

strideynet committed:
Update rfd/0162-machine-id-token-join-method-bot-instance.md

strideynet commented:
I've updated this to be marked implemented, @zmb3 / @timothyb89 PTAL so we can merge this RFD down.

timothyb89 committed:
Add details on bot authentication process as it was implemented

We discovered a third authentication option during implementation, so
add details on what we ended up using.
